### PR TITLE
Add metadata to metrics

### PR DESCRIPTION
**Pull request by hlbarber**
**Motivation**

Closes https://github.com/metrics-rs/metrics/issues/370

**Changes**

- Add `&Metadata` argument to the `Recorder::register_*` methods
- Add `target: *` and `Level::*` arguments to macros.
    - `target` defaults to `module_path!()`
    - `Level` defaults to `Level::INFO`.

**TODO**

- [ ] Add documentation
- [ ] Add `info_counter!`, `warn_counter!`, etc
- [ ] Improve test coverage

### CODE REVIEW

**Comment by hlbarber** [comment]
Added this in a draft state to garner comments before polishing it up.

**Comment by tobz** [comment]
@hlbarber High-level thoughts: this is pretty straightforward, but the macros should probably look more like the `tracing` macros.

In and of itself, `Metadata` feels as straightforward as I imagined it to be, but the lack of macro examples and the copied approach of how the `describe_*` macros handle detecting the unit parameter have solidified my thoughts around the macros needing to handle this better and in a way that's entirely compatible with the current macros. This is separate from actually providing level-prefixed variants of the macros, which I'm still a little hesitant about.

Essentially, what I'd want to see is:

- `target` and `level` require a struct field notation in the macro arguments (like `tracing`)
- metric name is always the first argument
- parse the remaining arguments as either the "fixed function" `target` or `level` fields, or as labels (label parsing logic stays as-is)
- cannot specify `target` or `level` _after_ we've parsed something label-like (set a flag once we see the first non-`target`/`level` arg, basically)

Or with examples:

```
register_counter!("metric_name");
register_counter!("metric_name", "label_a" => "...");
register_counter!("metric_name", target: "overridden");
register_counter!("metric_name", level: Level::TRACE, &some_vec_of_labels);
register_counter!("metric_name", target: "overridden", level: Level::Info, "service" => "foo");
```

Ultimately, seeing this PR, I'd say that you could keep going with everything else on your list but hold off on the `info_counter!` stuff. If you're willing to take a crack at the procedural macro changes I've proposed above, that'd be good to. Otherwise, I would likely submit those as changes directly to your PR in the next week or two.

**Comment by hlbarber** [comment]
[`tracing::span`](https://docs.rs/tracing/latest/tracing/macro.span.html) and [`tracing::event`](https://docs.rs/tracing/latest/tracing/macro.event.html), in order, accepts:

- Optional target: `target: {target}`
- Required level: `Level::{LEVEL}`
- Required name: `"{name}"` (for `span!`)

This implementation, in order, accepts:

- Optional target: `target: {target}`
- Optional level: `Level::{LEVEL}` (for backwards compatibility)
- Required metrics name: `"{name}"`
- ...

```rust
register_counter!("metric_name");
register_counter!("metric_name", "label_a" => "...");
register_counter!(target: "overridden", "metric_name");
register_counter!(Level::TRACE, "metric_name", &some_vec_of_labels);
register_counter!(target: "overridden", Level::Info, "metric_name", "service" => "foo");
```

> target and level require a struct field notation in the macro arguments

Agree, I can change `level: Level::{LEVEL}` _or_ make `Level::{LEVEL}` required (like `tracing`) - either of these will avoid the "unit parameter approach" (which I agree should perhaps be avoided). I think the struct field notation helps for optional fields.

_Side note: Perhaps the unit parameter should be struct field notation too? Maybe this allows these macros to be done declaratively rather than procedurally._

> metric name is always the first argument

Can do, but this seems like a departure `tracing` where

```rust
macro_rules! span {
    (target: $target:expr, parent: $parent:expr, $lvl:expr, $name:expr) => { ... };
    (target: $target:expr, parent: $parent:expr, $lvl:expr, $name:expr, $($fields:tt)*) => { ... };
```

**Comment by tobz** [comment]
> I can change `level: Level::{LEVEL}` _or_ make `Level::{LEVEL}` required (like `tracing`) - either of these will avoid the "unit parameter approach" (which I agree should perhaps be avoided). I think the struct field notation helps for optional fields.

Right, I want `level` to use the struct field notation because it it's optional.

> _Side note: Perhaps the unit parameter should be struct field notation too? Maybe this allows these macros to be done declaratively rather than procedurally._

The hard-coded bits of the `describe_*!` macros are going away with the rework to metric "attributes" -- https://github.com/metrics-rs/metrics/issues/337#issuecomment-1511547332 -- so I would just ignore that for the moment.

> Can do, but this seems like a departure `tracing` where

Fair point. I wasn't thinking we'd necessarily mirror everything about how the `tracing` macros do it, but I suppose all named optionals having to come first would make parsing easier, especially if we attempted to rewrite the macros to be declarative.

I'm also not against doing that either -- proc macros are heavy -- but it might need to be staged as an improvement that comes after this and metrics attributes, because those two items get rid of the grossest parts of the procedural macros, which is the wishy-washy handling is specific positional arguments (i.e. is this arg a `Unit`? let's check some hardcoded paths).

File: metrics/src/metadata.rs
Comment (tobz) [inline metrics/src/metadata.rs:27]: Hmm, why exactly is module path optional but not target?

Seems like we could swap that around, and then have `target()` return `self.target.as_ref().unwrap_or(self.module_path)` or something like that.

File: metrics/src/metadata.rs
Comment (tobz) [inline metrics/src/metadata.rs:1]: Obviously this was in place for the draft aspect, but just noting it that we'll need to drop this and add docs for all the new public types here.

**Review by tobz — CHANGES_REQUESTED** [review]
After sitting down and going over this, I feel a lot better about the nature of the change and what we're adding.

Left two small comments.

File: metrics/src/metadata.rs
Comment (hlbarber) [reply metrics/src/metadata.rs:27]: I've just imitated `tracing`s behavior here.

https://github.com/tokio-rs/tracing/blob/e603c2a254d157a25a7a1fbfd4da46ad7e05f555/tracing-core/src/metadata.rs#L68-L70

> The name of the Rust module where the span occurred, or `None` if this could not be determined.

Admittedly I don't know _when_ this cannot be determined given that `module_path!` returns a `&'static str`.

I will do some investigation about the reason why `tracing::Metadata` is setup like this and then get back to you.

Comment (hlbarber) [reply metrics/src/metadata.rs:42]: Summarizing my brief discussion with the `tracing` maintainers:

- `module_path` should _always_ be populated when the `Metadata` is constructed from the `tracing` macros.
- It is represented as an `Option`:
    1. To maintain compatibility with `log` whose [`module_path`](https://docs.rs/log/latest/log/struct.Record.html#method.module_path) is also optional.
    1. To interop with events emitted in contexts where `module_path` might not be readily accessible (outside of Rust).

We don't need to be compatible with `log` however it might be nice to enable the use case (ii).

File: metrics-exporter-prometheus/src/builder.rs
Comment (tobz) [reply metrics-exporter-prometheus/src/builder.rs:583]: Eh, it's hard to see the value at this point, given that there's only three parameters needed. I guess, maybe, if we added something like `metrics::default_metadata!()`?

Avoiding needing to type out the `module_path!()` macro in the right place would be one thing, but I don't know if the complexity would be worth it to make it support a builder pattern (i.e. use the default level but set a custom target) over just having people call `Metadata::new` themselves... or providing builder pattern methods on `Metadata`.

Comment (hlbarber) [reply metrics-exporter-prometheus/src/builder.rs:583]: > providing builder pattern methods on Metadata

Regarding this, would you prefer it? I've been quite conservative with what we've added to `Metadata`. The `Metadata::new` obviously ties us to the current set of fields, if you anticipate perhaps adding something like [`Metadata::file`](https://docs.rs/tracing/latest/tracing/struct.Metadata.html#method.file) in the future then `Metadata::builder` is probably preferable?

Comment (tobz) [reply metrics-exporter-prometheus/src/builder.rs:583]: Yeah, let's just wait.

While the builder methods would make it easier to write a declarative macro (since we could just collect each build method fragment and smoosh them together at the end) there's still a little bit tied up in setting/specifying the defaults for the current field and it's not clear to me what the sharp edges, if any, will be with the current approach... so I think we probably just need to use it first.

File: metrics/src/metadata.rs
Comment (tobz) [inline metrics/src/metadata.rs:33]: Looks like this needs to be changed to something like `[metrics macros][metrics_macros]` and then add `[metrics_macros]: https://docs.rs/metrics/latest/metrics/#macros` at the bottom of the overall doc comment.

I don't think there's a way to do intradoc links with _just_ a crate name, but it's also cleaner/more intuitive, I think, to link the macros section of the `metrics` crate since that's where people will be consuming them from.

**Review by tobz — CHANGES_REQUESTED** [review]
Left one little doc comment thing, and it looks like this needs to have `cargo fmt` run on it, but overall very close. 👍🏻

**Comment by tobz** [comment]
I'm going to try and take a look, in a few days, at just finalizing the remaining requested changes myself since it's so close.

**Comment by hlbarber** [comment]
"Rust / feature-check" is throwing a ` note: /usr/bin/ld: final link failed: No space left on device`, any suggestions?

**Review by tobz — APPROVED** [review]
LGTM.

The feature check CI step has been finicky lately with "out of disk space" errors cropping up a lot. I just check this stuff manually before cutting releases and it's usually fine, so just gonna ignore it for now.

**Pull request merged**
